### PR TITLE
Added preClose hook

### PR DESCRIPTION
**Pull request by mcollina**
Adds a `preClose` hook to remove the monkey patching in https://github.com/fastify/fastify-websocket/blob/1d1f8fa1f295413a82a1a38dbe2e23d7deb9e193/index.js#L156-L168.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Reference/Hooks.md [hunk 1]
```diff
@@ -434,6 +437,30 @@ fastify.addHook('onClose', async (instance) => {
...
 ```
 
+### preClose
+<a id="pre-close"></a>
+
+Triggered when `fastify.close()` is invoked to stop the server, before all in-flight
+HTTP requests has been completed.
+It is useful when [plugins](./Plugins.md) has set up some state attached
+to the HTTP server that would prevent the server to close.
+_It is unlilely you would need to use this hook_,
```
Comment (Fdawgs) [inline docs/Reference/Hooks.md:447]:
```suggestion
_It is unlikely you will need to use this hook_,
```

**Review by Fdawgs — CHANGES_REQUESTED** [review]

File: docs/Reference/Hooks.md [hunk 2]
```diff
@@ -414,9 +416,10 @@ fastify.addHook('onReady', async function () {
 ### onClose
 <a id="on-close"></a>
 
-Triggered when `fastify.close()` is invoked to stop the server. It is useful
-when [plugins](./Plugins.md) need a "shutdown" event, for example, to close an
-open connection to a database.
+Triggered when `fastify.close()` is invoked to stop the server, after all in-flight
+HTTP requests has been completed.
+It is useful when [plugins](./Plugins.md) need a "shutdown" event, for example,
+to close an open connection to a database.
```
Comment (jsumners) [inline docs/Reference/Hooks.md:422]:
```suggestion
Triggered when `fastify.close()` is invoked to stop the server, after all in-flight
HTTP requests have been completed.
It is useful when [plugins](./Plugins.md) need a "shutdown" event, for example,
to close an open connection to a database.
```

File: test/close.test.js [hunk 3]
```diff
@@ -495,3 +495,50 @@ test('shutsdown while keep-alive connections are active (non-async, custom)', t
     })
   })
 })
+
+test('preClose callback', t => {
```
Comment (climba03003) [inline test/close.test.js:499]: Can we test with multiple `preClose` hook and mixed `callback` with `async-await` to ensure the execution order.

```js
test('preClose execution order', t => {
  t.plan(5)
  async function sleep(ms) {
    return new Promise((resolve) => {
      setTimeout(resolve, ms)
    })
  }
  const fastify = Fastify()
  const order = []
  fastify.addHook('onClose', onClose)
  function onClose (instance, done) {
    t.type(fastify, instance)
    t.same(order, [1, 2, 3])
    done()
  }

  fastify.addHook('preClose', (done) => {
    setTimeout(function () {
      order.push(1)
      done()
    }, 200)
  })

  fastify.addHook('preClose', async () => {
    await sleep(100)
    order.push(2)
  })

  fastify.addHook('preClose', (done) => {
    setTimeout(function () {
      order.push(3)
      done()
    }, 100)
  })

  fastify.listen({ port: 0 }, err => {
    t.error(err)

    fastify.close((err) => {
      t.error(err)
      t.ok('close callback')
    })
  })
})
```

Comment (mcollina) [reply test/close.test.js:499]: added

**Comment by mcollina** [comment]
@Fdawgs @climba03003 PTAL

**Review by climba03003 — APPROVED** [review]

**Review by Fdawgs — APPROVED** [review]
Lgtm. Do the Typescript types need to be added as part of this PR, or does it not matter?

**Comment by climba03003** [comment]
> Do the Typescript types need to be added as part of this PR, or does it not matter?

For me, it is better to include on the same PR.

However, missing types is not a valid reason for blocking the PR from merging.
Anyone interested can submit another PR just for the types.

**Review by metcoder95 — APPROVED** [review]

**Comment by mcollina** [comment]
Added types, PTAL

**Pull request merged**